api/v1alpha1/backend_types.go

```diff
@@ -0,0 +1,108 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// KindBackend is the name of the Backend kind.
+	KindBackend = "Backend"
+)
+
+type ProtocolType string
+
+const (
+	// ProtocolTypeTCP defines the TCP address protocol.
+	ProtocolTypeTCP ProtocolType = "TCP"
+	// ProtocolTypeUDP defines the UDP address protocol.
+	ProtocolTypeUDP ProtocolType = "UDP"
+)
+
+type ApplicationProtocolType string
+
+const (
+	// ApplicationProtocolTypeH2C defines the HTTP/2 prior knowledge application protocol.
+	ApplicationProtocolTypeH2C ApplicationProtocolType = "H2C"
+	// ApplicationProtocolTypeWS defines the WebSocket over HTTP protocol.
+	ApplicationProtocolTypeWS ApplicationProtocolType = "WS"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=envoy-gateway,shortName=be
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[?(@.type=="Accepted")].reason`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+//
+// Backend allows the user to configure the behavior of the connection
+// between the Envoy Proxy listener and the backend service.
+type Backend struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec defines the desired state of Backend.
+	Spec BackendSpec `json:"spec"`
+
+	// status defines the current status of Backend.
+	Status BackendStatus `json:"status,omitempty"`
+}
+
+// BackendAddress describes are backend address, which is can be either a TCP/UDP socket or a Unix Domain Socket
+// corresponding to Envoy's Address: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-address
+//
+// +kubebuilder:validation:XValidation:rule="(has(self.socketAddress) || has(self.unixDomainSocketAddress))",message="one of socketAddress or unixDomainSocketAddress must be specified"
+// +kubebuilder:validation:XValidation:rule="(has(self.socketAddress) && !has(self.unixDomainSocketAddress)) || (!has(self.socketAddress) && has(self.unixDomainSocketAddress))",message="only one of socketAddress or unixDomainSocketAddress can be specified"
+type BackendAddress struct {
+	// Name is the unique name of the backend address
+	Name string `json:"name,omitempty"`
+
```

Comment:
should we add a top level `Type` - FQDN, IPv4, IPv6, UDS ?

Reply:
UDS is distinguished from other types with a dedicated struct, since it doesn't follow the `host:port` pattern anyway. We can add IPv4/IPv6/FQDN type field. The value that I see for such a type is simplification of validation (only need to check against the expected type). However, it creates a more verbose API. I'm OK with both. WDYT?

Reply:
thoughts on ?
```
	FQDN *SocketAddress `json:"fqdn,omitempty"`
	
	IP *SocketAddress `json:"ip,omitempty"`

	// Unix defines the unix domain socket path
	//
	// +optional
	Unix *UnixDomainSocketAddress `json:"unix,omitempty"`
```

Reply:
Agree. When ahead and created dedicated structs to make validation a bit easier. For a yaml perspective, FQDN/IP look the same.